synapse/storage/util/id_generators.py

```diff
@@ -251,30 +258,82 @@ def __init__(
 
         self._sequence_gen = PostgresSequenceGenerator(sequence_name)
 
+        # This goes and fills out the above state from the database.
+        self._load_current_ids(db_conn, table, instance_column, id_column)
+
     def _load_current_ids(
         self, db_conn, table: str, instance_column: str, id_column: str
-    ) -> Dict[str, int]:
-        # If positive stream aggregate via MAX. For negative stream use MIN
-        # *and* negate the result to get a positive number.
-        sql = """
-            SELECT %(instance)s, %(agg)s(%(id)s) FROM %(table)s
-            GROUP BY %(instance)s
-        """ % {
-            "instance": instance_column,
-            "id": id_column,
-            "table": table,
-            "agg": "MAX" if self._positive else "-MIN",
-        }
-
+    ):
         cur = db_conn.cursor()
-        cur.execute(sql)
 
-        # `cur` is an iterable over returned rows, which are 2-tuples.
-        current_positions = dict(cur)
+        # Load the current positions of all writers for the stream.
+        if self._writers:
+            sql = """
+                SELECT instance_name, stream_id FROM stream_positions
+                WHERE stream_name = ?
+            """
+            sql = self._db.engine.convert_param_style(sql)
 
-        cur.close()
+            cur.execute(sql, (self._stream_name,))
+
+            self._current_positions = {
+                instance: stream_id * self._return_factor
+                for instance, stream_id in cur
+                if instance in self._writers
```

Comment:
By excluding only the known writers here could there be a dataloss situation when removing a writer, if that writer is the minimum position?

If you have writers:
* A at pos 15
* B at pos 17
* C at pos 12

And then remove "C", `_current_positions` will be {A: 15, B: 17} and `min_stream_id` will be 15. I'm not sure if this is OK or not?

It seems you also don't want to include writers that you no longer care about or else you can end up in a situation where you have an old writer always saying it is far behind, which I'm guessing this if-statement checks against?

Reply:
As per discussion elsewhere: this is OK because if `C` has been removed from the list it means its been turned off, and so we know its up to date.

It's worth noting that these positions aren't really "where `C` has gotten up to in the rooms", but instead is more "where `C` is currently persisting events". I.e. we're not worried that if we remove `C` from the deployment then `A` or `B` will start persisting at position 12, as either `C` finished persisting a row at 12 *or* the request gets retried and `A` or `B` will persist it with a *new* position.